master/static/srv/proto_get_trial_workloads.sql

```diff
@@ -0,0 +1,62 @@
+WITH validations_vt AS (
+  SELECT row_to_json(r1) AS validation,
+    prior_batches_processed
+  FROM (
+      SELECT 'STATE_' || v.state as state,
+        v.start_time,
+        v.end_time,
+        s.num_batches,
+        s.prior_batches_processed,
+        v.metrics->'num_inputs' as num_inputs,
+        v.metrics->'validation_metrics' as metrics
+      FROM validations v
+        INNER JOIN steps s ON v.trial_id = s.trial_id
+        AND v.step_id = s.id
+      WHERE v.trial_id = $1
+    ) AS r1
+),
+trainings_vt AS (
+  SELECT row_to_json(r1) AS training,
+    prior_batches_processed
+  FROM (
+      SELECT s.start_time,
+        s.end_time,
+        'STATE_' || s.state as state,
+        s.num_batches,
+        s.prior_batches_processed,
+        s.metrics->'avg_metrics' as metrics,
+        s.metrics->'num_inputs' as num_inputs
+      FROM steps s
+      WHERE s.trial_id = $1
+    ) AS r1
+),
+checkpoints_vt AS (
+  SELECT row_to_json(r1) AS checkpoint,
+    prior_batches_processed
+  FROM (
+      SELECT 'STATE_' || c.state as state,
+        c.start_time,
+        c.end_time,
+        c.uuid,
+        s.num_batches,
+        s.prior_batches_processed,
+        c.resources
+      FROM checkpoints c
+        INNER JOIN steps s ON c.trial_id = s.trial_id
+        AND c.step_id = s.id
+      WHERE c.trial_id = $1
+    ) AS r1
+)
+SELECT v.validation,
+  t.training,
+  c.checkpoint
+FROM trainings_vt t
+  FULL JOIN checkpoints_vt c ON false
+  FULL JOIN validations_vt v ON false
```

Comment:
I looked more into doing this with `UNION ALL` but still couldn't get it to deser into protobuf correctly (even wrapped the results in `json_build_object` to get what I thought it was expecting. So this is fine with me.

Reply:
Can you post a query plan before we merge though, to look at?

Reply:
```
"QUERY PLAN"
"Sort  (cost=51.27..51.27 rows=2 width=104)"
"  Sort Key: (COALESCE(t.start_time, COALESCE(v.start_time, c.start_time)))"
"  CTE validations_vt"
"    ->  Nested Loop  (cost=4.57..19.81 rows=1 width=40)"
"          Join Filter: (v_1.step_id = s.id)"
"          ->  Index Scan using validations_trial_step_unique on validations v_1  (cost=0.28..8.29 rows=1 width=128)"
"                Index Cond: (trial_id = 20)"
"          ->  Bitmap Heap Scan on steps s  (cost=4.29..11.48 rows=2 width=16)"
"                Recheck Cond: (trial_id = 20)"
"                ->  Bitmap Index Scan on steps_pkey  (cost=0.00..4.29 rows=2 width=0)"
"                      Index Cond: (trial_id = 20)"
"  CTE trainings_vt"
"    ->  Bitmap Heap Scan on steps s_1  (cost=4.29..11.51 rows=2 width=40)"
"          Recheck Cond: (trial_id = 20)"
"          ->  Bitmap Index Scan on steps_pkey  (cost=0.00..4.29 rows=2 width=0)"
"                Index Cond: (trial_id = 20)"
"  CTE checkpoints_vt"
"    ->  Nested Loop  (cost=4.57..19.81 rows=1 width=40)"
"          Join Filter: (c_1.step_id = s_2.id)"
"          ->  Index Scan using checkpoints_trial_step_unique on checkpoints c_1  (cost=0.28..8.29 rows=1 width=82)"
"                Index Cond: (trial_id = 20)"
"          ->  Bitmap Heap Scan on steps s_2  (cost=4.29..11.48 rows=2 width=16)"
"                Recheck Cond: (trial_id = 20)"
"                ->  Bitmap Index Scan on steps_pkey  (cost=0.00..4.29 rows=2 width=0)"
"                      Index Cond: (trial_id = 20)"
"  ->  Merge Full Join  (cost=0.00..0.13 rows=2 width=104)"
"        Join Filter: false"
"        ->  Merge Full Join  (cost=0.00..0.09 rows=2 width=80)"
"              Join Filter: false"
"              ->  CTE Scan on trainings_vt t  (cost=0.00..0.04 rows=2 width=40)"
"              ->  Materialize  (cost=0.00..0.02 rows=1 width=40)"
"                    ->  CTE Scan on checkpoints_vt c  (cost=0.00..0.02 rows=1 width=40)"
"        ->  Materialize  (cost=0.00..0.02 rows=1 width=40)"
"              ->  CTE Scan on validations_vt v  (cost=0.00..0.02 rows=1 width=40)"

```

Reply:
Can we swap out the sort by time for a sort by step id? That part is worse than I thought.

Reply:
Since that didn't make much of a different, maybe just go back to time. I explain'd the current "TrialDetailsRaw" and it's worse, so this is most likely fine.

Reply:
(sorry)

Reply:
The only thing I can thing of that would be better maybe is pulling one more CTE out for that "Bitmap Heap Scan" That happens in all three CTEs. Something like
```explain WITH full_steps_for_trial AS (
    SELECT
        -- training and common details
        s.start_time as training_start_time,
        s.end_time as training_end_time,
        'STATE_' || s.state as training_state,
        s.num_batches,
        s.prior_batches_processed,
        s.id as step_id,
        s.metrics->'avg_metrics' as training_metrics,
        s.metrics->'num_inputs' as training_num_inputs,
        -- validation details
        v.start_time validation_start_time,
        v.end_time validation_end_time,
        'STATE_' || v.state as validation_state,
        v.metrics->'validation_metrics' as validation_metrics,
        v.metrics->'num_inputs' as validation_num_inputs,
        -- checkpoint details
        'STATE_' || c.state as checkpoint_state,
        c.start_time as checkpoint_start_time,
        c.end_time as checkpoint_end_time,
        c.uuid as checkpoint_uuid,
        c.resources as checkpoint_resources
    FROM steps s
    LEFT JOIN validations v on s.id = v.step_id
    LEFT JOIN checkpoints c on s.id = c.step_id
    WHERE s.trial_id = 1
),
     trainings_vt AS (
     SELECT row_to_json(r1) AS training,
            step_id
     FROM (
              SELECT f.training_start_time,
                     f.training_end_time,
                     'STATE_' || f.training_state as state,
                     f.num_batches,
                     f.prior_batches_processed,
                     f.step_id,
                     f.training_metrics as metrics,
                     f.training_num_inputs as num_inputs
              FROM full_steps_for_trial f
          ) AS r1
 ),
    validations_vt AS (
    SELECT row_to_json(r1) AS validation,
           step_id
    FROM (
             SELECT 'STATE_' || f.validation_state as state,
                    f.validation_start_time,
                    f.validation_end_time,
                    f.num_batches,
                    f.prior_batches_processed,
                    f.step_id,
                    f.validation_num_inputs as num_inputs,
                    f.validation_metrics as metrics
             FROM full_steps_for_trial f
         ) AS r1
),
     checkpoints_vt AS (
     SELECT row_to_json(r1) AS checkpoint,
            step_id
     FROM (
              SELECT 'STATE_' || f.checkpoint_state as state,
                     f.checkpoint_start_time,
                     f.checkpoint_end_time,
                     f.checkpoint_uuid,
                     f.step_id,
                     f.num_batches,
                     f.prior_batches_processed,
                     f.checkpoint_resources as resources
              FROM full_steps_for_trial f
          ) AS r1
 )
SELECT v.validation::jsonb - 'step_id' AS validation,
       t.training::jsonb - 'step_id' AS training,
       c.checkpoint::jsonb - 'step_id' AS checkpoint
FROM trainings_vt t
 FULL JOIN checkpoints_vt c ON false
 FULL JOIN validations_vt v ON false
ORDER BY coalesce(t.step_id, coalesce(v.step_id, c.step_id)) ASC
```
Gets a query plan like
```
Sort  (cost=80.19..80.26 rows=31 width=100)
"  Sort Key: (COALESCE(t.step_id, COALESCE(v.step_id, c.step_id)))"
  CTE full_steps_for_trial
    ->  Hash Right Join  (cost=27.09..50.51 rows=31 width=332)
          Hash Cond: (v_1.step_id = s.id)
          ->  Seq Scan on validations v_1  (cost=0.00..18.80 rows=880 width=56)
          ->  Hash  (cost=27.01..27.01 rows=7 width=132)
                ->  Hash Right Join  (cost=12.69..27.01 rows=7 width=132)
                      Hash Cond: (c_1.step_id = s.id)
                      ->  Seq Scan on checkpoints c_1  (cost=0.00..13.40 rows=340 width=72)
                      ->  Hash  (cost=12.64..12.64 rows=4 width=64)
                            ->  Bitmap Heap Scan on steps s  (cost=4.18..12.64 rows=4 width=64)
                                  Recheck Cond: (trial_id = 1)
                                  ->  Bitmap Index Scan on steps_pkey  (cost=0.00..4.18 rows=4 width=0)
                                        Index Cond: (trial_id = 1)
  CTE trainings_vt
    ->  CTE Scan on full_steps_for_trial f  (cost=0.00..0.78 rows=31 width=36)
  CTE validations_vt
    ->  CTE Scan on full_steps_for_trial f_1  (cost=0.00..0.78 rows=31 width=36)
  CTE checkpoints_vt
    ->  CTE Scan on full_steps_for_trial f_2  (cost=0.00..0.78 rows=31 width=36)
  ->  Merge Full Join  (cost=0.00..26.58 rows=31 width=100)
        Join Filter: false
        ->  Merge Full Join  (cost=0.00..13.25 rows=31 width=72)
              Join Filter: false
              ->  CTE Scan on trainings_vt t  (cost=0.00..0.62 rows=31 width=36)
              ->  Materialize  (cost=0.00..0.70 rows=31 width=36)
                    ->  CTE Scan on checkpoints_vt c  (cost=0.00..0.62 rows=31 width=36)
        ->  Materialize  (cost=0.00..0.70 rows=31 width=36)
              ->  CTE Scan on validations_vt v  (cost=0.00..0.62 rows=31 width=36)
```
Which is maybe marginally better, but I don't think it's huge. But this endpoint does get polled frequently.

Reply:
Lmk what you think.

Reply:
I don't think it makes much a difference though since we're inner joining from vals/checkpoints to steps, so nevermind again lol (I guess we save selecting training rows with vals or checkpoints twice but there aren't many of those anyway).

Reply:
the CTEs this one make seems to be bigger in size, could that be more susceptible to the database going OOM, if that happens? 

about lowering the 
` ->  Bitmap Heap Scan on steps s  (cost=4.18..12.64 rows=4 width=64)`

and SQL in general I don't really know as much as you do so your call

Reply:
I'm fine with it either way, they're so close it's not going to make or break us. But we should make sure to scale test it at release party.

Reply:
alright I'll keep it as is then, it'll be easy to swap out and test given if we run into a scaling issue. Also, if we have a metric endpoint in the future this workload list could certainly be paginated and that should help with the load.